src/pycsou/opt/solver/lsmr.py

```diff
@@ -0,0 +1,332 @@
+        exist_ok: bool = False,
+        writeback_rate: typ.Optional[int] = None,
+        verbosity: int = 1,
+        show_progress: bool = True,
+        log_var: pyct.VarName = ("x",),
+    ):
+        super().__init__(
+            folder=folder,
+            exist_ok=exist_ok,
+            writeback_rate=writeback_rate,
+            verbosity=verbosity,
+            show_progress=show_progress,
+            log_var=log_var,
+        )
+
+        self._A = A
+        self._atol = atol
+        self._btol = btol
+        self._ctol = (1.0 / conlim) if (conlim > 0) else 0.0
+        self._iter_lim = (min(A.shape[0], A.shape[1])) if (iter_lim is None) else iter_lim
+        self._damp = damp
+        self._itn = 0
+        self._normb = None
+
+    @pycrt.enforce_precision(i=["b", "x0"], allow_None=True)
+    def m_init(
+        self,
+        b: pyct.NDArray,
+        x0: typ.Optional[pyct.NDArray] = None,
+    ):
+        xp = pycu.get_array_module(b)
+
+        b = xp.atleast_1d(b)
+        if b.ndim > 1:
+            b = b.squeeze()
+        m, n = self._A.shape
+
+        u = b
+        normb = xp.linalg.norm(b)
+
+        if x0 is None:
+            x = xp.zeros(n)
+            beta = normb.copy()
+        else:
+            x = x0.copy()
+            u -= self._A.apply(x)
+            beta = xp.linalg.norm(u)
+
+        if beta > 0:
+            u *= 1 / beta
+            v = self._A.T.apply(u)
+            alpha = xp.linalg.norm(v)
+        else:
+            v = xp.zeros_like(x)
+            alpha = 0
+
+        if alpha > 0:
+            v *= 1 / alpha
+
+        mst = self._mstate
+
+        normar = alpha * beta
+        if normar == 0:
+            mst["trivial"] = True
+            if b.ndim == 1:
+                mst["x"] = xp.zeros(n)
+            elif b.ndim == 2:
+                mst["x"] = xp.zeros((b.shape[0], n))
+            return
+
+        mst["x"], mst["u"], mst["v"], mst["alpha"], mst["trivial"] = x, u, v, alpha, False
+
+        # Initialize variables for 1st iteration:
+        mst["zetabar"], mst["alphabar"] = alpha * beta, alpha
+        mst["rho"] = mst["rhobar"] = mst["cbar"] = 1
+        mst["sbar"] = 0
+        mst["h"], mst["hbar"] = v.copy(), xp.zeros_like(x)
+
+        # Initialize variables for estimation of ||r||:
+        mst["betadd"] = beta
+        mst["rhodold"] = 1
+        mst["betad"] = mst["tautildeold"] = mst["thetatilde"] = mst["zeta"] = mst["d"] = 0
+
+        # Initialize variables for estimation of ||A|| and cond(A):
+        mst["normA2"] = alpha**2
+        mst["maxrbar"], mst["minrbar"] = 0, 1e100
+        mst["normA"], mst["condA"] = xp.sqrt(mst["normA2"]), 1
+        mst["normx"], mst["normr"], mst["normar"] = 0, beta, normar
+        self._normb = normb
+
+        # Initialize variables for testing stopping rules:
+        mst["test1"], mst["test2"], mst["test3"] = 1, alpha / beta, None
+
+    @staticmethod
+    def _sym_ortho(a, b, xp):
```

Comment:
`a` and `b` can be NDArrays

Reply:
Fixed that by using xp.vectorize function. This is the only function that exists both in numpy, dask.array and cupy. I am not sure if it's the fastest solution to make this function to be vectorized, but it works with that. There is also frompyfunc function, but it doesn't exist in cupy, that's why I used vectorize